src/analysis/processing/qgsalgorithmalignrasters.cpp

```diff
@@ -0,0 +1,182 @@
+}
+
+QgsAlignRastersAlgorithm *QgsAlignRastersAlgorithm::createInstance() const
+{
+  return new QgsAlignRastersAlgorithm();
+}
+
+void QgsAlignRastersAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterAlignRasterLayers( QStringLiteral( "LAYERS" ), QObject::tr( "Input layers" ) ) );
+  addParameter( new QgsProcessingParameterRasterLayer( QStringLiteral( "REFERENCE_LAYER" ), QStringLiteral( "Reference layer" ) ) );
+
+  addParameter( new QgsProcessingParameterCrs( QStringLiteral( "CRS" ), QObject::tr( "CRS" ), QVariant(), true ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "CELL_SIZE_X" ), QObject::tr( "Cell size X" ), QgsProcessingParameterNumber::Double, QVariant(), true, 1e-9 ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "CELL_SIZE_Y" ), QObject::tr( "Cell size Y" ), QgsProcessingParameterNumber::Double, QVariant(), true, 1e-9 ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "GRID_OFFSET_X" ), QObject::tr( "Grid offset X" ), QgsProcessingParameterNumber::Double, QVariant(), true, 1e-9 ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "GRID_OFFSET_Y" ), QObject::tr( "Grid offset Y" ), QgsProcessingParameterNumber::Double, QVariant(), true, 1e-9 ) );
+  addParameter( new QgsProcessingParameterExtent( QStringLiteral( "EXTENT" ), QObject::tr( "Clip to extent" ), QVariant(), true ) );
+
+  addOutput( new QgsProcessingOutputMultipleLayers( QStringLiteral( "OUTPUT_LAYERS" ), QObject::tr( "Aligned rasters" ) ) );
+}
+
+struct QgsAlignRasterProgress : public QgsAlignRaster::ProgressHandler
+{
+    explicit QgsAlignRasterProgress( QgsFeedback *feedback ) : mFeedback( feedback ) {}
+    bool progress( double complete ) override
+    {
+      mFeedback->setProgress( complete * 100 );
+      return true;
+    }
+
+  protected:
+    QgsFeedback *mFeedback = nullptr;
+};
+
+
+QVariantMap QgsAlignRastersAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QgsRasterLayer *rasterLayer = parameterAsRasterLayer( parameters, QStringLiteral( "REFERENCE_LAYER" ), context );
+
+  if ( !rasterLayer )
+    throw QgsProcessingException( invalidRasterError( parameters, QStringLiteral( "REFERENCE_LAYER" ) ) );
+
+  const QVariant layersVariant = parameters.value( parameterDefinition( QStringLiteral( "LAYERS" ) )->name() );
+  const QList<QgsAlignRasterData::RasterItem> items = QgsProcessingParameterAlignRasterLayers::parameterAsItems( layersVariant, context );
+  bool hasReference = false;
+  QStringList outputLayers;
+  outputLayers.reserve( items.size() );
+  for ( const QgsAlignRasterData::RasterItem &item : items )
+  {
+    if ( !hasReference && item.inputFilename == rasterLayer->source() )
+    {
+      hasReference = true;
+    }
+
+    outputLayers << item.outputFilename;
+  }
+
+  if ( !hasReference )
```

Comment:
This has always confused me in the old align rasters tool -- why must the reference layer be included in the list of layers to align? If the other values aren't being changed (ie the cell size etc) then the reference layer will be untouched by the operation, and it seems odd that it has to be included in this list...

Reply:
Well, personally I have the same feeling. We discussed this with @wonder-sk and he said that this was intentional to make sure that all rasters, including the reference one are in the same format, resampling, etc.

Reply:
Hm, if it's up for debate them I'm -1 to that constraint :laughing:

Reply:
While this was originally intentional, I do not have a strong opinion about it - so it could be removed if you prefer it that way :slightly_smiling_face: